mockgcp/common/httptogrpc/service.go

```diff
@@ -54,49 +110,67 @@ func (m *grpcMux) AddService(client any) {
 	}
 
 	// Use the protobuf types to find the FileDescriptor, from there we can find the services
-	var matchingServices []protoreflect.ServiceDescriptor
+	matchingServices := make(map[string]protoreflect.ServiceDescriptor)
 	for _, protoType := range discoveredProtobufTypes {
 		msg := reflect.New(protoType).Elem().Interface()
 		md := msg.(proto.Message).ProtoReflect().Descriptor()
-		fd := md.ParentFile()
 
-		services := fd.Services()
-		for i := range services.Len() {
-			service := services.Get(i)
+		protoregistry.GlobalFiles.RangeFilesByPackage(md.ParentFile().FullName(), func(fd protoreflect.FileDescriptor) bool {
+			services := fd.Services()
+			for i := range services.Len() {
+				service := services.Get(i)
 
-			isMatch := true
-			methods := service.Methods()
-			for j := range methods.Len() {
-				method := methods.Get(j)
-				if !goTypeMethodNames[string(method.Name())] {
-					isMatch = false
-					break
+				if matchingServices[string(service.FullName())] != nil {
+					// Already found
+					continue
 				}
-			}
 
-			// TODO: Is there a better way to match this?
-			if isMatch {
-				matchingServices = append(matchingServices, service)
+				// We match by looking at the method names.  There may be a better way to do this, but we haven't found one yet!
+				isMatch := true
+
+				protoMethods := service.Methods()
+				protoMethodNames := make(map[string]bool)
+				for j := range protoMethods.Len() {
+					method := protoMethods.Get(j)
+					if !goTypeMethodNames[string(method.Name())] {
+						isMatch = false
+						break
+					}
+					protoMethodNames[string(method.Name())] = true
+				}
+				if !isMatch {
+					continue
+				}
+
+				for goTypeMethodName := range goTypeMethodNames {
+					if !protoMethodNames[goTypeMethodName] {
+						isMatch = false
+						break
+					}
+				}
+				if !isMatch {
+					continue
+				}
+
+				matchingServices[string(service.FullName())] = service
 			}
-		}
-		if len(matchingServices) > 0 {
-			break
-		}
+			return true
+		})
 	}
 
-	if len(matchingServices) == 0 {
-		klog.Fatalf("cannot match service for %v", goType.Name())
-	}
 	if len(matchingServices) > 1 {
-		klog.Fatalf("found multiple matching service for %v", goType.Name())
+		for k, v := range matchingServices {
```

Comment:
If `matchingServices` has > 1 entries, `klog.Infof` prints them.
It would be helpful to also print the `client` type name in the loop or in the final Fatalf message (which you do).

Reply:
The idea is that we print the client type name once, at the end.  Printing the client type every time was just TMI